1. Add login page illustration

WiNloSt opened the pull request:
Epic: #39246

### Description

This PR adds a setting for the new login page custom illustration. Only one of the four settings is added. The rest 3 settings are easy enough to add, so I'll add them in a subsequent PR to keep this PR size as minimal as possible.

### How to verify

1. Login as admin using an EE instance.
1. Go to Admin > Appearance > Conceal Metabase
1. Play around with the new login page illustration

### Demo
![image](https://github.com/metabase/metabase/assets/1937582/238e208f-1972-4fc2-a2a0-da492e5f1833)
![image](https://github.com/metabase/metabase/assets/1937582/7be24c2a-1362-48ab-a22d-816e639c80b8)



### Checklist

- [x] Tests have been added/updated to cover changes in this PR

WiNloSt committed:
Move Metabot setting to a new place

WiNloSt committed:
Add setting for login page illustration

WiNloSt committed:
fix UI styling

WiNloSt committed:
Fix not able to upload the same file twice

WiNloSt committed:
Update the admin > appearance > conceal metabase copy

WiNloSt committed:
Prepare the component for tests

WiNloSt committed:
Add some unit tests

WiNloSt committed:
Prevent settings not saved sometimes

After setting the value to/from `custom` one of the setting could be
failed to save, I suspect this happens because the race condition from
trying to save the setting values relatively at the same time.

WiNloSt committed:
Make unit tests easier to read

WiNloSt committed:
Add all unit test test cases

WiNloSt committed:
Add basic E2E test

WiNloSt committed:
Show fallback icon for broken images

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/whitelabel/components/ImageToggle/ImageToggle.tsx:
```diff
@@ -30,17 +30,13 @@ const ImageToggle = ({
       <ToggleContainer>
         <ToggleLabel htmlFor={toggleId}>{label}</ToggleLabel>
-        <Toggle
+        <Switch
```
This is unrelated to the main change (new settings), but it changes the Metabot toggle to use the new component from `metabase/ui`
![image](https://github.com/metabase/metabase/assets/1937582/57f93c76-0dab-4833-ba58-10753991ef77)

WiNloSt committed:
Make use of the proper component

WiNloSt committed:
Update the comment to affect the reality

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/whitelabel/components/LoginPageIllustrationDescription.js:
```diff
@@ -0,0 +1,26 @@
+        {t`Login page`}
+        <HoverCard position="top-start">
+          <HoverCard.Target>
+            <Icon aria-hidden name="info" />
```
`aria-hidden` makes sure we could use `*ByLabelText("Login page")` later on, though I'm not using that right now.

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/whitelabel/index.js:
```diff
@@ -211,6 +184,38 @@ if (hasPremiumFeature("whitelabel")) {
+            widget: MetabotToggleWidget,
+          },
+          {
+            key: "login-page-illustration",
```
It's easy enough to add 3 more settings once this PR is approved, so I want to make sure I get the foundation right first.

That's why this PR only contain one new setting.

WiNloSt commented on frontend/src/metabase/auth/components/AuthLayout/AuthLayout.tsx:
```diff
@@ -17,11 +16,17 @@ interface AuthLayoutProps {
+        <LayoutIllustration
+          data-testid="login-page-container"
+          backgroundImageSrc={loginPageIllustration}
+          isDefault={loginPageIllustration === "app/img/bridge.svg"}
```
I don't feel good about this. I want 2 pieces of information here.
1. Whether or not the src is the default option because we render it based on the brand color
2. The source image of the custom option.

This works for now, but maybe I'll change the selector so it returns an object that has `isDefault` and `src` properties. Any thought about that?

WiNloSt replied:
Fixed in 74349ee3110f4d4fd955348ab572f86685be87cb

WiNloSt commented:
@noahmoss I tagged you as a reviewer for the BE change (1 file) since GitHub suggested it 😄

vbenedetti commented:
I was reviewing the PR and bumped into a few glitches relative to the thumbnail preview and uploading of larger images (40MB for example). I recorded this loom with my findings: https://www.loom.com/share/3ec05b45caf944729c111b046f146694?sid=5d273098-e8b8-4d2e-b25b-4d82328a1141

WiNloSt committed:
Add an error message when file size is over 2MB

WiNloSt committed:
Show an error message when the file size > 2MB

WiNloSt committed:
Add the error message tests

WiNloSt committed:
Preserve the preview image aspect-ratio

WiNloSt committed:
Detect corrupted images

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/whitelabel/components/IllustrationWidget/IllustrationWidget.tsx:
```diff
@@ -0,0 +1,193 @@
+
+type IllustrationValue = "default" | "no-illustration" | "custom";
+
+export interface StringSetting {
```
I could use the name of the widget instead like this example https://github.com/metabase/metabase/blob/51f9caf975b7368b52649aefd8594b52e71ae0ac/enterprise/frontend/src/metabase-enterprise/whitelabel/components/LighthouseToggleWidget/types.ts#L1-L4

so we could use `IllustrationSetting` or something. The fact that I didn't want to use that because I don't really care what setting this is, in fact, I only care that the `setting.value` is a string so I named it that way.

For me if we were to name something like `IllustrationSetting` or `<insert actual setting name>Setting` I'd think it's the type of the whole setting instead. Something like
https://github.com/metabase/metabase/blob/51f9caf975b7368b52649aefd8594b52e71ae0ac/enterprise/frontend/src/metabase-enterprise/whitelabel/index.js#L207-L222

So... What do you think?

npretto replied:
The thing is that I read "StringSetting", but then the actual value is IllustrationValue, even though technically the underlaying type is a string (union)

Maybe to avoid having to come up with names, we could think about having a wrapper generic type, such as `SettingValue<T>` or `SettingValueWithDefault<T>` but that's something we'd have to chat about with team webapp, for this PR it's ok as it is now 👍

WiNloSt replied:
I like the generic solution. Could be a good option going forward, but agree let's not doing anything in this PR right now.

WiNloSt committed:
Fix failing unit tests

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/whitelabel/components/LoginPageIllustrationDescription.tsx:
```diff
@@ -0,0 +1,32 @@
+        {t`Login page`}
+        <HoverCard position="top-start">
+          <HoverCard.Target>
+            <Icon aria-hidden name="info" />
```
in what way exactly? Width or preview border or the select width?
1. Setting width: I think I made it to almost match the existing Metabot toggle.
2. Preview image border: Maybe I could add a bit of vertical padding to the preview image, so it doesn't feel too tight in there.
3. Select width: It's only narrower when the custom option is selected to accommodate for the file name or the `Remove uploaded image` description (which will displays when you navigate away and back to this setting again since we don't store the image name anyway, we couldn't display the image name except only when users are uploading them)

WiNloSt replied:
I'll see what I can do. Thanks for the clarification.

WiNloSt replied:
Welp, forgot to do this I'll include it in the next PR.

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/whitelabel/components/IllustrationWidget/IllustrationWidget.tsx:
```diff
@@ -0,0 +1,214 @@
+  errorMessageContainerId: string;
+};
+
+const MB = 1024 * 1024;
```
I like this. Thanks!

WiNloSt replied:
Fixed in e80a72c59c4da0147019b69e3c32d1ec5c3686da

WiNloSt committed:
Address feedback: Type

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/whitelabel/components/IllustrationWidget/IllustrationWidget.tsx:
```diff
@@ -0,0 +1,214 @@
+
+import { LighthouseImage, PreviewImage } from "./IllustrationWidget.styled";
+
+type IllustrationValue = "default" | "no-illustration" | "custom";
```
fixed in 7a1b445519ac8539ccf987e8665dc434b50f2533

WiNloSt committed:
Feedback: Create meaningful variables

WiNloSt committed:
Feedback: Make selector API more sensible

deniskaber reviewed (approved):
Added some final small comments.
Please fix them and then it is good by me!

WiNloSt commented on enterprise/frontend/src/metabase-enterprise/whitelabel/components/IllustrationWidget/IllustrationWidget.tsx:
```diff
@@ -0,0 +1,214 @@
+        }
+        setFileName(file.name);
+        // Setting 2 setting values at the same time could result in one of them not being saved
+        await onChange("custom");
```
> In some places you call this with `async`, sometimes without `async`. I think we should make it consistent.

Just for clarification, you mean sometimes I call `onChange` sometimes I call `await onChange` right? If that's the case then the comment above the `await onChange` calls already explains that. Normally I don't want to await things unnecessarily unless I have to, which is why I only have `await` in 2 places where we change 2 setting values in quick succession.

Maybe I could use a better structure like (quick idea)
```ts
function scheduleSettingChanges(settingPromises) {
  // do nothing when there is only one promise, because we don't need to wait when changing only 1 settings

  // await the first setting or both (I'd rather prefer to await only the first one to not waste time unnecessarily)
}
```

usage could look something like this
```ts
scheduleSettingChanges([onChange("no-illustration")])
...
scheduleSettingChanges([onChange("custom"), onChangeSetting(customIllustrationSetting, dataUri)])
```

what do you think?


> Also, we should either use strict typings for onChange, or move "default", "no-illustration", "custom" to be constants.

Totally agree, I think I missed this.

WiNloSt replied:
Fixed in 83360bfde95995f252df6ff676ec3629ef4fd2cc and 78142c20b0d666e22bd1fb4f2d891e3d0d2c257f


I ended up calling everything with `await` since creating another abstraction isn't worth it, it makes the code too complicated.

WiNloSt committed:
Prevent race condition

Makes sure we save the settings completely before making other actions
that also change settings otherwise, there could be race conditions.

WiNloSt committed:
Feedback: tighten up a type

WiNloSt committed:
Feedback: Fix inconsistency callback calls
